ipykernel/ipkernel.py

```diff
@@ -261,6 +270,42 @@ def do_complete(self, code, cursor_pos):
                 'metadata' : {},
                 'status' : 'ok'}
 
+    def _experimental_do_complete(self, code, cursor_pos):
+        """
+        Experimental completions from IPython, using Jedi. 
+        """
+        if cursor_pos is None:
+            cursor_pos = len(code)
+        with provisionalcompleter():
+            raw_completions = self.shell.Completer.completions(code, cursor_pos)
+            completions = list(rectify_completions(code, raw_completions))
+            
+            comps = []
+            for comp in completions:
+                comps.append(dict(
+                            start=comp.start,
+                            end=comp.end,
+                            text=comp.text,
+                            type=comp.type,
+                              ))
+
+        if completions:
+            s = completions[0].start
+            e = completions[0].end
+            matches = [c.text for c in completions]
+        else:
+            s = cursor_pos
+            e = cursor_pos
+            matches = []
+
+        return {'matches': matches,
+                'cursor_end': e,
+                'cursor_start': s,
+                'metadata': {'_experimental': comps},
```

Comment:
The only big question is how to encode this in the message data. I think this approach of putting this into the metadata makes sense. Maybe add a suffix to help a frontend to distinguish if there are multiple experimental extensions:

`_experimental_type_data`